### ISSUE TITLE
How does poetry handle a conflict of python versions?

### ISSUE THREAD

**Issue by stefanoborini** [python-poetry/poetry#3864]
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [documentation](https://python-poetry.org/docs/) and believe that my question is not covered.

## Issue

I have the following question: imagine that you create a lock, but poetry finds out that package foo 1.2.0 works only with python 2.7, and foo 1.3.0 works only with python 3.x.

From what I see of the lock, every package has only one [[package]] entry with one version of the package. 
Can it produce a lock that, when installed with python 2.7, creates an environment with foo 1.2.0, and when installed with python 3.x, creates an environment with foo 1.3.0?

If so, how would it look in the lock file?

**Comment by stefanoborini** [comment]
@finswimmer yes but I should have said, what happens when as a consequence of the resolution of the tree, this is found in a subdependency?

**Comment by finswimmer** [comment]
As long as the dependencies specify their need correct poetry will write the different versions for the different python versions to the lock file. Something like this:

```
[[package]]
name = "pandas"
version = "0.25.3"
description = "Powerful data structures for data analysis, time series, and statistics"
category = "main"
optional = false
python-versions = ">=3.5.3"

[package.dependencies]
numpy = ">=1.13.3"
python-dateutil = ">=2.6.1"
pytz = ">=2017.2"

[package.extras]
test = ["pytest (>=4.0.2)", "pytest-xdist", "hypothesis (>=3.58)"]

[[package]]
name = "pandas"
version = "1.2.3"
description = "Powerful data structures for data analysis, time series, and statistics"
category = "main"
optional = false
python-versions = ">=3.7.1"

[package.dependencies]
numpy = ">=1.16.5"
python-dateutil = ">=2.7.3"
pytz = ">=2017.3"

[package.extras]
test = ["pytest (>=5.0.1)", "pytest-xdist", "hypothesis (>=3.58)"]
```

**Comment by stefanoborini** [comment]
@finswimmer ok so both resolution trees are present basically (with their equal dependencies made unique), and when it's time to install, the resolution tree that satisfies the constraints of the current python version will be selected.

Sorry for the bit technical questions. What I am doing is reimplementing poetry for R, so I am facing more or less the same problems and I am trying to understand what kind of solutions you picked.
